### PR TITLE
[Kernel] Start SWA/chunked KV-tile loop at first allowed key (#44575)

### LINKED ISSUE #44575: [Performance]: Triton unified attention: sliding-window / chunked KV-tile loop iterates an extra tile when the window start is not tile-aligned
## Summary

In `triton_unified_attention.py`, the sliding-window / chunked-attention KV-tile loop starts at a floor-rounded tile boundary (`first_allowed_key // TILE_SIZE`) rather than at the window's first allowed key. The leading slots before `first_allowed_key` in that first tile are masked out by `seq_mask`, but they still consume part of the tile, so a window of `W` keys can span one more `TILE_SIZE` tile than the minimal `ceil(W / TILE_SIZE)`.

## Detail

`compute_tile_loop_bounds` (`triton_attention_helpers.py`):

```python
tile_start = tl.maximum(0, first_allowed_key // TILE_SIZE)  # floor-rounded
```

paired with the main loop:

```python
for j in range(loop_lo, loop_hi):
    seq_offset = j * TILE_SIZE + offs_t                     # starts on a tile boundary
```

A window of `W = last_allowed_key - first_allowed_key + 1` keys starting at residue `r = first_allowed_key % TILE_SIZE` spans `ceil((r + W) / TILE_SIZE)` tiles with the floor-rounded start, versus the minimal `ceil(W / TILE_SIZE)` if iteration begins at `first_allowed_key` - up to one extra tile iteration (KV load + MMA + softmax merge). No tile is fully masked; the first tile is partially masked at its front, and that misalignment is what pushes the window's tail into the extra tile.

Example (`TILE_SIZE = 32`, window `[31, 46]`, `W = 16`, `r = 31`): floor-rounded start iterates tiles `{0, 1}` (the window's keys split 1 + 15); a window-aligned start iterates 1.

## Proposed fix

Have `compute_tile_loop_bounds` return a base offset `tile_base` and iterate `seq_offset = tile_base + j * TILE_SIZE + offs_t`, beginning exactly at `first_allowed_key`. The window then occupies the minimal `ceil(W / TILE_SIZE)` tiles.

- Non-SWA paths and the 3D-segmented SWA path are unchanged (`tile_base = 0`; the 3D path reasons in absolute tile indices).
- The tensor-descriptor (`USE_TD`) path is also unchanged (`tile_base = 0`): its KV load fetches a whole tile from one block and relies on `BLOCK_SIZE % TILE_SIZE == 0`, which an arbitrary base would break.
- Only the 2D pointer sliding-window / chunked path shifts its base; its per-slot `physical_block_idx = seq_offset // BLOCK_SIZE` load already handles a tile straddling two blocks.

## Secondary effect

Removing the leading masked slots also makes the online-softmax reduction order independent of `r` (the window's offset modulo `TILE_SIZE`), so identical Q against an identical `(K, V)` window yields byte-identical output across batch shapes (previously off by bf16 ULPs). A regression test covering this is included.

### Report of performance regression

_No response_

### Misc discussion on performance

_No response_

### Your current environment (if you think it is necessary)

_No response_

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by AjAnubolu**
## Summary

Fixes #44575. The sliding-window/chunked KV-tile loop started at a floor-rounded
tile (`max(0, first_allowed_key // TILE_SIZE)`), so a window of `W` keys spanned
up to one extra `TILE_SIZE` tile beyond the minimal `ceil(W / TILE_SIZE)`.

`compute_tile_loop_bounds` now returns a `tile_base` so the loop starts exactly at
`first_allowed_key`. It's non-zero only for the 2D-pointer path; a new `USE_TD`
constexpr keeps it 0 for the tensor-descriptor and 3D paths. `tile_end` is
unchanged, so the iteration count never grows. The `diffkv` variant shares the
helper and is updated too.

## Tests

Added a residue-sweep `seq_lens` row, validated against the PyTorch reference on a B200:

```
test_triton_unified_attention.py -k seq_lens2        576 passed
test_triton_unified_attention_diffkv.py -k seq_lens2  48 passed
```

Pre-existing `use_td` failures reproduce on clean main (#44850); untouched here.
No open PR targets #44575.

### CODE REVIEW

**Comment by almogtavor** [comment]
Hey @AjAnubolu, thanks! I think this PR can be closed. I've already implemented it at https://github.com/vllm-project/vllm/pull/44584 as well, but I closed it, so we'll first see some more interest in this idea. The reason it's not a trivial win is that the tiles may indeed not be aligned, but it may still be optimal because of an alignment with the actual physical memory structure. My experiments didn't really prove anything, so I only kept the issue (https://github.com/vllm-project/vllm/issues/44575)
